File: 068_CountingBits.py
"""
Topic: Dynamic Programming / Bit Tricks
Problem: Counting Bits
Description: For each i in [0..n], count number of 1 bits.
Time: O(n)
Space: O(n)
"""

from typing import List


def countBits(n: int) -> List[int]:
    res = [0] * (n + 1)
    for i in range(1, n+1):
        res[i] = res[i & (i-1)] + 1
    return res


if __name__ == '__main__':
    print(countBits(2))  # [0,1,1]
    print(countBits(5))  # [0,1,1,2,1,2]

File: 068_CountingBits.cpp
// Topic: Dynamic Programming / Bit Tricks
// Counting Bits: compute number of set bits for all numbers 0..n using dp relation
// Time Complexity: O(n)
// Space Complexity: O(n)

#include <bits/stdc++.h>
using namespace std;

vector<int> countBits(int n)
{
    vector<int> res(n + 1);
    for (int i = 1; i <= n; i++) res[i] = res[i & (i - 1)] + 1;
    return res;
}

int main()
{
    auto r = countBits(5);
    for (int x : r) cout << x << " ";
    cout << "\n";  // 0 1 1 2 1 2
    r = countBits(0);
    for (int x : r) cout << x << " ";
    cout << "\n";  // 0
    return 0;
}
